### ISSUE TITLE
Issue when try to run on local Tomcat 8 server

### ISSUE THREAD

**Issue by wangzhidong** [watson-developer-cloud/questions-with-classifier-ega#65]
I downloaded the code to local. First run mvn install for fremawork-ega and generated the jar, and then try to run the questions-with-classifier-ega-war on local Tomcat 8. But I'm always getting following exception,, could you please advise, thanks so much.

INFO: Command line argument: -Dfile.encoding=UTF-8
Mar 07, 2016 10:00:57 PM org.apache.catalina.core.AprLifecycleListener lifecycleEvent
INFO: The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /Users/tony/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.
Mar 07, 2016 10:00:57 PM org.apache.coyote.AbstractProtocol init
INFO: Initializing ProtocolHandler ["http-nio-8080"]
Mar 07, 2016 10:00:57 PM org.apache.tomcat.util.net.NioSelectorPool getSharedSelector
INFO: Using a shared selector for servlet write/read
Mar 07, 2016 10:00:57 PM org.apache.coyote.AbstractProtocol init
INFO: Initializing ProtocolHandler ["ajp-nio-8009"]
Mar 07, 2016 10:00:57 PM org.apache.tomcat.util.net.NioSelectorPool getSharedSelector
INFO: Using a shared selector for servlet write/read
Mar 07, 2016 10:00:57 PM org.apache.catalina.startup.Catalina load
INFO: Initialization processed in 663 ms
Mar 07, 2016 10:00:57 PM org.apache.catalina.core.StandardService startInternal
INFO: Starting service Catalina
Mar 07, 2016 10:00:57 PM org.apache.catalina.core.StandardEngine startInternal
INFO: Starting Servlet Engine: Apache Tomcat/8.0.30
Mar 07, 2016 10:00:58 PM org.apache.catalina.core.StandardContext filterStart
SEVERE: Exception starting filter CORSFilter
java.lang.ClassNotFoundException: com.ibm.watson.app.common.util.rest.CORSFilter
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1333)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1167)
	at org.apache.catalina.core.DefaultInstanceManager.loadClass(DefaultInstanceManager.java:520)
	at org.apache.catalina.core.DefaultInstanceManager.loadClassMaybePrivileged(DefaultInstanceManager.java:501)
	at org.apache.catalina.core.DefaultInstanceManager.newInstance(DefaultInstanceManager.java:120)
	at org.apache.catalina.core.ApplicationFilterConfig.getFilter(ApplicationFilterConfig.java:258)
	at org.apache.catalina.core.ApplicationFilterConfig.<init>(ApplicationFilterConfig.java:105)
	at org.apache.catalina.core.StandardContext.filterStart(StandardContext.java:4659)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5281)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1408)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1398)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

Mar 07, 2016 10:00:58 PM org.apache.catalina.core.StandardContext startInternal
SEVERE: One or more Filters failed to start. Full details will be found in the appropriate container log file
Mar 07, 2016 10:00:58 PM org.apache.catalina.core.StandardContext startInternal
SEVERE: Context [] startup failed due to previous errors

**Comment by tdalbo92** [comment]
@wangzhidong Thanks for opening this.  

We'll try to work through this, but be advised that we've never supported nor advocated using Tomcat.  All of our development and deployment was done with WebSphere Liberty 8.5.5.7.  I wrote a custom Filter to enable CORS for our API, and that's bound in web.xml.  If you're sure running `mvn install` in framework-ega/ is successful, and running `mvn install` in questions-with-classifier/ is also successful, then I suspect there's some type of discrepancy between how Liberty and Tomcat do classpath lookups when resolving Filters.  

I'll download Tomcat and see if I run into the same issue.  Would you be able to use Liberty instead of Tomcat for whatever purpose you're building this app? 

Thanks,
Tom

**Comment by wangzhidong** [comment]
@tdalbo92 , Thanks for your quick feedback.
I changed to WebSphere Liberty server, and now I'm able to start the application, but when I try to access the index page, getting following error, could you please help to advise:

15:18:41.059 [Default Executor-thread-11] WARN  com.ibm.watson.app.common.services.impl.BluemixServicesBinder - AQWEGA12000W - No VCAP_SERVICES JSON configuration was available
[ERROR   ] CWWJP9992E: javax.persistence.PersistenceException: CWWJP0013E: The server cannot locate the java:comp/env/jdbc/db data source for the com.ibm.watson.app.qaclassifier.db persistence unit because it has encountered the following exception: javax.naming.NamingException: CWNEN1001E: The object referenced by the java:comp/env/jdbc/db JNDI name could not be instantiated. If the reference name maps to a JNDI name in the deployment descriptor bindings for the application performing the JNDI lookup, make sure that the JNDI name mapping in the deployment descriptor binding is correct. If the JNDI name mapping is correct, make sure the target resource can be resolved with the specified name relative to the default initial context. [Root exception is com.ibm.wsspi.injectionengine.InjectionException: CWNEN0030E: The server was unable to obtain an object instance for the java:comp/env/jdbc/db reference.  The exception message was: CWNEN1003E: The server was unable to find the jdbc/questions-with-classifier-ega-db binding with the javax.sql.DataSource type for the java:comp/env/jdbc/db reference.].
[ERROR   ] SRVE0271E: Uncaught init() exception created by servlet [Classifier JAX-RS Servlet] in application [questions-with-classifier-ega-war]: javax.servlet.ServletException: Resource class com.ibm.watson.app.qaclassifier.ClassifierApplicationImpl can not be instantiated due to InvocationTargetException
	at com.ibm.ws.jaxrs20.server.LibertyJaxRsServerFactoryBean.createSingletonInstance(LibertyJaxRsServerFactoryBean.java:220)
	at [internal classes]
	at com.ibm.watson.app.common.util.rest.CORSFilter.doFilter(CORSFilter.java:38)
	at com.ibm.ws.webcontainer.filter.FilterInstanceWrapper.doFilter(FilterInstanceWrapper.java:207)
	at [internal classes]

**Comment by tdalbo92** [comment]
@wangzhidong Hey, I have an update for you.

I was able to download Tomcat and partially get the application to deploy.  However, when I tried to make any REST calls to the application they were returning a 404.  I suspect further configuration in web.xml is needed.  I also had to modify a few things that were WebSphere Liberty specific, like the default servlet binding.  I suspect the 404 is due to classpath issues, but I didn't investigate further.  I'll just say that it's not advised to run on Tomcat, and your mileage may vary.  If you do get it working, we would very much appreciate a Pull Request with the working code so everybody can benefit from it.

Since you're running Liberty locally, but not in Bluemix, you'll need to provide a VCAP_SERVICES object in your environment.  You can look at server.env as an example of how this environment variable should look.

**Comment by wangzhidong** [comment]
@tdalbo92 , thanks Tom, could you please provide a  VCAP_SERVICES object example in your project. and attach the file you used and the path where to put this file, thanks so much.
Of course, please hide your confidential  information.

**Comment by tdalbo92** [comment]
As I mentioned previously, we provide one of these examples in the code.

https://github.com/watson-developer-cloud/questions-with-classifier-ega/blob/master/questions-with-classifier-ega-test/src/it/resources/server.bluemix.env

Just replace the appropriate credentials with yours, provided by the Bluemix service instance when you create it, and place it here.  It might be easier if you create an empty application in Bluemix, create all the service instances and bind them to that application, and then use the VCAP_SERVICES in Bluemix as a model to create your own locally.

That file needs to be named `server.env` placed in `usr/servers/servername` of your Liberty installation, where `servername` is the name of your server.  It's usually `defaultServer`, but double check that name first.

**Comment by wangzhidong** [comment]
Hi Tom, thanks for your clarification.
I put a server.env to my local usr/servers/servername of Liberty installation and replaced the content with my credentials, but I'm still getting the error. I guess need to config the database also like natural_language_classifier. 
But I don't know where to put this information. Could you please kindly advise again? Thanks so much.

**Comment by wangzhidong** [comment]
To be frankly, I just want to deploy this application to my local environment and debug the code of  REST calls.

**Comment by tdalbo92** [comment]
Oh!  Haha, I missed the second error in that initial message you sent and was only focusing on the VCAP_SERVICES binding. 

`CWNEN1003E: The server was unable to find the jdbc/questions-with-classifier-ega-db binding with the javax.sql.DataSource type for the java:comp/env/jdbc/db reference.].`

This means it didn't find the db binding with that name.  If you look in questions-with-classifier-ega-test/pom.xml, we're setting that specific name `jdbc/questions-with-classifier-ega-db` as a property.  Then it's passed into the system properties for Liberty to use when starting up.  

Long story short, you'll have to look at `questions-with-classifier-ega-test/pom.xml` and `src/it/resources/server.xml` to see how to configure this variable, and get it running in your local liberty server.

**Comment by tdalbo92** [comment]
We initially were developing and deploying locally with the Liberty plugin in Eclipse, but quickly found out that it was far too burdensome and fragile for a dev environment.  It is possible, but it's been so long that I don't remember exactly how to configure it.

**Comment by wangzhidong** [comment]
I add the following configuration in server.xml for my local Liberty server.
	<jdbcDriver id="derby-driver">
		<library name="derby-lib">
			<fileset dir="C:\\EGA\\questions-with-classifier-ega\\questions-with-classifier-ega-test\\target/lib/derby" includes="derby.jar, derbyclient.jar" />
		</library>
	</jdbcDriver>
	<dataSource id="derby-db" jdbcDriverRef="derby-driver"
		jndiName="jdbc/questions-with-classifier-ega-db" transactional="true">
		
		<properties.derby.embedded createDatabase="create"
			databaseName="SQLDB" />
	</dataSource>

I'm able to deploy the app successfully, but I'm getting another error:
Internal Exception: java.sql.SQLSyntaxErrorException: Table/View 'ANSWERS' does not exist.
Error Code: 30000
Call: SELECT CLASS, TEXT, TYPE, CANONICAL_QUESTION FROM ANSWERS
Query: ReadAllQuery(name="Answers.getAllAnswers" referenceClass=AnswerEntity sql="SELECT CLASS, TEXT, TYPE, CANONICAL_QUESTION FROM ANSWERS")


Do you have the DDLs that to generate the tables which used in application? Or is there any other ways to create all the tables?
